lib/ace/ext/papeeria/katex-previewer.coffee

```diff
@@ -50,43 +54,142 @@ define((require, exports, module) ->
+
+      getPopoverPosition: (row) ->
+        secondRowSelector = "div.ace_scroller > div > div.ace_layer.ace_text-layer > div:nth-child(2)"
+        jqSecondRow = jqEditorContainer.find(secondRowSelector)
+        secondRowPosition = jqSecondRow.position()
+        pxRowHeight = jqSecondRow.height()
+        relativeRow = row + 1 - ch.getTopmostRowNumber()
+        top = "#{secondRowPosition.top + pxRowHeight * (relativeRow + 1)}px"
+
+        left = jqEditorContainer.position().left
+
+        return {
+          top: top
+          left: left
+        }
 
-    onLoaded = ->
-      try
-        content = katex.renderToString(
-          editor.getSelectedText(),
+      getCurrentFormula: ->
+        katex.renderToString(
+          ch.getWholeEquation(ch.curStart, ch.curEnd),
           {displayMode: true}
         )
-      catch e
-        content = e
-      finally
-        popoverHandler.show($("#formula"), content)
-        editor.on("changeSelection", callbackHidePopover)
-        editor.session.on("changeScrollTop", callbackHidePopover)
-        editor.session.on("changeScrollLeft", callbackHidePopover)
-        return
 
-    callback = (editor) -> createPopover(editor)
+      initPopover: ->
+        {row: cursorRow} = editor.getCursorPosition()
+        [ch.curStart, ch.curEnd] = ch.getEquationRange(cursorRow)
+        popoverPosition = ch.getPopoverPosition(ch.curEnd)
+        try
+          content = ch.getCurrentFormula()
+        catch e
+          content = e
+        finally
+          popoverHandler.show(jqFormula(), content, popoverPosition)
+
+      updatePopover: ->
+        try
+          content = ch.getCurrentFormula()
+        catch e
+          content = e
+        finally
+          popoverHandler.setContent(jqFormula(), content)
+
+      delayedUpdatePopover: ->
+        if ch.currentDelayedUpdateId?
+          clearTimeout(ch.currentDelayedUpdateId)
+        ch.currentDelayedUpdateId = setTimeout((-> ch.updatePopover(); currentDelayedUpdateId = null), 1000)
+
+      updatePosition: ->
+        popoverHandler.setPosition(jqFormula(), ch.getPopoverPosition(ch.curEnd))
+
+      handleCurrentContext: ->
+        currentContext = editor.session.getContext(editor.getCursorPosition().row)
+        if ch.prevContext != "equation" and currentContext == "equation"
+          ch.contextPreviewExists = true
+          if not katex?
+            initKaTeX(ch.initPopover)
+          else
+            ch.initPopover()
+          editor.on("change", ch.delayedUpdatePopover)
+          editor.session.on("changeScrollTop", ch.updatePosition)
+        else if ch.prevContext == "equation" and currentContext != "equation"
+          ch.contextPreviewExists = false
+          editor.off("change", ch.delayedUpdatePopover)
+          editor.session.off("changeScrollTop", ch.updatePosition)
+          popoverHandler.destroy(jqFormula())
 
-    destroyPopover = -> popoverHandler.hide($("#formula"))
+        ch.prevContext = currentContext
```

Comment:
isn't it enough to check `contextPreviewExists` only? If it is `false` and you're in equation, you show popover, if it is `true` and you're not in equation, you hide it